utils/binstat/binstat_external_test.go

```diff
@@ -0,0 +1,232 @@
+	zlog = unittest.Logger()
+
+	// delete any files hanging around from previous test run
+	{
+		command := "ls -al ./binstat.test.pid-*.binstat.txt ./*gomaxprocs*.pprof.txt ; rm -f ./binstat.test.pid-*.binstat.txt ./*gomaxprocs*.pprof.txt"
+		out, err := exec.Command("bash", "-c", command).Output()
+		require.NoError(t, err)
+		zlog.Debug().Msgf("test: output of command: %s\n%s", command, out)
+	}
+
+	// run the test; loops of several tries running groups of go-routines
+	for loop := 0; loop < loops; loop++ {
+		gomaxprocs := 8
+		if 0 == loop {
+			gomaxprocs = 1
+		}
+		p := binstat.EnterTime(fmt.Sprintf("loop-%d", loop), "")
+		for try := 0; try < tries; try++ {
+			zlog.Debug().Msgf("test: loop=%d try=%d; running 6 identical functions with gomaxprocs=%d", loop, try+1, gomaxprocs)
+			run(t, loop, try, gomaxprocs)
+		}
+		binstat.Leave(p)
+	}
+
+	// output a table of results similar to this
+	/*
+		- binstat------- pprof---------
+		- try1 try2 try3 try1 try2 try3
+		- 0.29 0.30 0.29 0.05 0.03 0.05 // f1() seconds; loop=0 gomaxprocs=1
+		- 0.35 0.30 0.35 0.07 0.06 0.06 // f2() seconds; loop=0 gomaxprocs=1
+		- 0.28 0.33 0.28 0.06 0.06 0.06 // f3() seconds; loop=0 gomaxprocs=1
+		- 0.31 0.28 0.31 0.05 0.06 0.06 // f4() seconds; loop=0 gomaxprocs=1
+		- 0.27 0.28 0.27 0.05 0.05 0.05 // f5() seconds; loop=0 gomaxprocs=1
+		- 0.38 0.38 0.39 0.06 0.05 0.06 // f6() seconds; loop=0 gomaxprocs=1
+		- binstat------- pprof---------
+		- try1 try2 try3 try1 try2 try3
+		- 0.07 0.07 0.07 0.05 0.03 0.07 // f1() seconds; loop=1 gomaxprocs=8
+		- 0.07 0.07 0.07 0.05 0.04 0.03 // f2() seconds; loop=1 gomaxprocs=8
+		- 0.07 0.07 0.07 0.04 0.07 0.07 // f3() seconds; loop=1 gomaxprocs=8
+		- 0.07 0.07 0.07 0.05 0.02 0.08 // f4() seconds; loop=1 gomaxprocs=8
+		- 0.07 0.07 0.07 0.09 0.06 0.07 // f5() seconds; loop=1 gomaxprocs=8
+		- 0.07 0.07 0.07 0.04 0.10 0.03 // f6() seconds; loop=1 gomaxprocs=8
+	*/
+	for loop := 0; loop < loops; loop++ {
+		zlog.Debug().Msg("test: binstat------- pprof---------")
+		l1 := "test:"
+		for r := 0; r < 2; r++ {
+			for try := 0; try < tries; try++ {
+				l1 = l1 + fmt.Sprintf(" try%d", try+1)
+			}
+		}
+		zlog.Debug().Msg(l1)
+		gomaxprocs := 8
+		if 0 == loop {
+			gomaxprocs = 1
+		}
+		for i := 0; i < funcs; i++ {
+			l2 := "test:"
+			for mech := 0; mech < mechs; mech++ {
+				for try := 0; try < tries; try++ {
+					l2 = l2 + fmt.Sprintf(" %s", el[loop][try][mech][i])
+				}
+			}
+			l2 = l2 + fmt.Sprintf(" // f%d() seconds; loop=%d gomaxprocs=%d", i+1, loop, gomaxprocs)
+			zlog.Debug().Msg(l2)
+		}
+	}
+
+	// tell binstat to close down and write its stats file one last time
+	binstat.Fin()
+
+	// cat and sort binstat stats file
+	{
+		command := "ls -al ./binstat.test.pid-*.binstat.txt ; cat ./binstat.test.pid-*.binstat.txt | sort --version-sort"
+		out, err := exec.Command("bash", "-c", command).Output()
+		require.NoError(t, err)
+		zlog.Debug().Msgf("test: output of command: %s\n%s", command, out)
```

Comment:
@yhassanzadeh13 thanks for the comment! Yes, I'm aware of that format, but the logging is purely used for debugging binstat itself (and therefore not intended for general consumption outside of debugging binstat), and is designed to give more human readable logs due to alignment (and at first glance it seems like zlog helper methods are not designed for this kind of carefully formatted output?), rather than more readable source code.

Reply:
I think going with the suggested pattern is more readable and maintainable over time.